asapdiscovery-data/asapdiscovery/data/schema_v2/fragalysis.py

```diff
@@ -0,0 +1,117 @@
+import os
+from pathlib import Path
+from typing import List
+
+import pandas
+from asapdiscovery.data.schema_v2.complex import Complex
+from asapdiscovery.data.schema_v2.schema_base import ContainerAbstractBase
+from pydantic import Field, root_validator
+
+
+class FragalysisFactory(ContainerAbstractBase):
+    """
+    Schema for a loading a Fragalysis dump. The directoory given by parent_dir should
+    contain (at a minimum):
+     * a metadata.csv file with the columns "crystal_name" and "alternate_name"
+     * an aligned/ subdirectory, containing a subdirectory for each entry in
+       metadata.csv, with these subdirectories containing the bound PDB files
+    """
+
+    parent_dir: Path = Field(
+        description="Top level directory of the Fragalysis database."
+    )
+    complexes: list[Complex] = Field(
+        [], description="Complex objects in the Fragalysis database.", repr=False
+    )
+
+    def __len__(self):
+        return len(self.complexes)
+
+    @classmethod
+    def from_dir(
+        cls,
+        parent_dir: str | Path,
+        xtal_col="crystal_name",
+        compound_col="alternate_name",
+    ) -> "FragalysisFactory":
+        """
+        Build a FragalysisFactory from a Fragalysis directory.
+
+        Parameters
+        ----------
+        parent_dir : str | Path
+            Top-level directory of the Fragalysis database
+        xtal_col : str, default="crystal_name"
+            Name of the column in metadata.csv giving the crystal names. Defaults to the
+            Fragalysis value. The values in this col MUST match the directories in
+            the aligned/ subdirectory
+        compound_col : str, default="alternate_name"
+            Name of the column in metadata.csv giving the compound names. Defaults to
+            the Fragalysis value
+
+        Returns
+        -------
+        FragalysisFactory
+        """
+        parent_dir = Path(parent_dir)
+        try:
+            df = pandas.read_csv(parent_dir / "metadata.csv")
+        except FileNotFoundError as e:
+            raise FileNotFoundError("No metadata.csv file found in parent_dir.") from e
+
+        if len(df) == 0:
+            raise ValueError("metadata.csv file is empty.")
+
+        if (xtal_col not in df.columns) or (compound_col not in df.columns):
+            raise ValueError(
+                "metadata.csv file must contain a crystal name column and a "
+                "compound name column."
+            )
+
+        # Dict mapping crystal name to compound name
```

Comment:
I might try and see if we can use more pandas ops here and below.

Reply:
this ended up being a great idea for helping to make sure that Factories loaded from the same directory have the complexes in the same order (previously relied on the `os.listdir`, which I think is not the same every time it's called)